newclient/client.go

```diff
@@ -0,0 +1,245 @@
+func New(options ...clientOption) (*Client, error) {
+	// We start with a client with bare zero-valued fields
+	c := &Client{}
+
+	// Then we modify it according to user-provided options...
+	for _, f := range options {
+		err := f(c)
+		if err != nil {
+			return c, err
+		}
+	}
+
+	// ... and check if the result is valid
+	if err := validate(c); err != nil {
+		return c, err
+	}
+
+	// Finally, we add just a sprinkle of defaults and a new Client is born!
+	return setDefaults(c), nil
+}
+
+// The WithAppengineURL function allows to specify an
+// AppEngine URL different from the standard one (e.g. http://localhost:4000).
+// This is not recommendend in production.
+func WithAppengineURL(appEngineURL string) clientOption {
+	return func(c *Client) error {
+		appengine, err := url.Parse(appEngineURL)
+		if err != nil {
+			return err
+		}
+		c.appEngineURL = appengine
+		return nil
+	}
+}
+
+// The WithHousekeepingURL function allows to specify an
+// Housekeeping URL different from the standard one (e.g. http://localhost:4001).
+// This is not recommendend in production.
+func WithHousekeepingURL(housekeepingURL string) clientOption {
+	return func(c *Client) error {
+		housekeeping, err := url.Parse(housekeepingURL)
+		if err != nil {
+			return err
+		}
+		c.housekeepingURL = housekeeping
+		return nil
+	}
+}
+
+// The WithPairingURL function allows to specify an
+// Pairing URL different from the standard one (e.g. http://localhost:4002).
+// This is not recommendend in production.
+func WithPairingURL(pairingURL string) clientOption {
+	return func(c *Client) error {
+		// check that it's a valid URL
+		pairing, err := url.Parse(pairingURL)
+		if err != nil {
+			return err
+		}
+		c.pairingURL = pairing
+		return nil
+	}
+}
+
+// The WithRealmManagementURL function allows to specify an
+// RealmManagement URL different from the standard one (e.g. http://localhost:4003).
+// This is not recommendend in production.
+func WithRealmManagementURL(realmManagementURL string) clientOption {
+	return func(c *Client) error {
+		realmManagement, err := url.Parse(realmManagementURL)
+		if err != nil {
+			return err
+		}
+		c.realmManagementURL = realmManagement
+		return nil
+	}
+}
+
+// The WithBaseURL function allows to specify the Astarte
+// base URL (e.g. api.your-astarte.org)
+func WithBaseURL(baseURL string) clientOption {
+	return func(c *Client) error {
+		base, err := url.Parse(baseURL)
+		if err != nil {
+			return err
+		}
+		c.baseURL = base
+
+		return nil
+	}
+}
+
+// The WithHTTPClient function allows to specify an httpClient
+// with custom options, e.g. different timeout, or skipTLSVerify
+func WithHTTPClient(httpClient *http.Client) clientOption {
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// The WithToken function allows to specify a JWT
+// token that the client will use to interact with Astarte.
+func WithToken(token string) clientOption {
```

Comment:
Do we want to call it `WithToken` or `WithJWT`? Since `astartectl` uses `gen-jwt` maybe the second one would be more coherent with the naming we're using around.